plugin-server/src/cdp/cdp-consumers.ts

```diff
@@ -124,17 +153,107 @@ abstract class CdpConsumerBase {
                         }
                     })
                 )
+            },
+        })
+    }
 
-                await Promise.all(
-                    messagesToProduce.map((x) =>
-                        this.kafkaProducer!.produce({
-                            topic: x.topic,
-                            value: Buffer.from(JSON.stringify(x.value)),
-                            key: x.key,
-                            waitForAck: true,
+    protected async executeAsyncResponses(
+        asyncResponses: HogFunctionInvocationAsyncResponse[]
+    ): Promise<HogFunctionInvocationResult[]> {
+        return await runInstrumentedFunction({
+            statsKey: `cdpConsumer.handleEachBatch.executeAsyncResponses`,
+            func: async () => {
+                this.hogWatcher.currentObservations.observeAsyncFunctionResponses(asyncResponses)
+                // Filter for blocked functions
+                asyncResponses = asyncResponses.filter((item) => {
+                    const functionState = this.hogWatcher.getFunctionState(item.hogFunctionId)
+
+                    if (functionState === HogWatcherState.overflowed) {
+                        // TODO: _Maybe_ report to AppMetrics 2 when it is ready
+                        this.messagesToProduce.push({
+                            topic: KAFKA_CDP_OVERFLOW,
+                            value: {
+                                source: 'hog_function_callback',
+                                payload: item,
+                            },
+                            key: item.id,
+                        })
+                        return false
+                    }
+                    if (functionState > HogWatcherState.disabledForPeriod) {
+                        // TODO: Report to AppMetrics 2 when it is ready
+                        return false
+                    }
+                    return true
+                })
```

Comment:
Nits: I think this would be easier to follow as a simple `for of` loop that pushes to either `this.messagesToProduce` or `filteredAsyncResponses` depending on the condition. I always find it counter intuitive when "pure" operations like `map`, `filter`, etc have side effects 🤷

Reply:
True. Thats a good idea

Reply:
Done in the follow up PR